node/forest_libp2p/src/rpc/blocksync_message.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use forest_blocks::BlockHeader;
+use forest_cid::Cid;
+use forest_encoding::{
+    de::{self, Deserialize, Deserializer},
+    ser::{self, Serialize, Serializer},
+};
+use forest_message::{SignedMessage, UnsignedMessage};
+
+/// The payload that gets sent to another node to request for blocks and messages. It get DagCBOR serialized before sending over the wire.
+#[derive(Clone, Debug, PartialEq)]
+pub struct BlockSyncRequest {
+    /// The tipset to start sync from
+    pub start: Vec<Cid>,
+    /// The amount of epochs to sync by
+    pub request_len: u64,
+    /// 1 = Block only, 2 = Messages only, 3 = Blocks and Messages
+    pub options: u64,
+}
+
+impl Serialize for BlockSyncRequest {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        (&self.start, &self.request_len, &self.options).serialize(serializer)
+    }
+}
+impl<'de> Deserialize<'de> for BlockSyncRequest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let (start, request_len, options) = Deserialize::deserialize(deserializer)?;
+        Ok(BlockSyncRequest {
+            start,
+            request_len,
+            options,
+        })
+    }
+}
+
+/// The response to a BlockSync request.
+#[derive(Clone, Debug, PartialEq)]
+pub struct BlockSyncResponse {
+    /// The tipsets requested
+    pub chain: Vec<TipSetBundle>,
+    /// Error code
+    pub status: u64,
+    pub message: String,
+}
+
+impl Serialize for BlockSyncResponse {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        (&self.chain, &self.status, &self.message).serialize(serializer)
+    }
+}
+impl<'de> Deserialize<'de> for BlockSyncResponse {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let (chain, status, message) = Deserialize::deserialize(deserializer)?;
+        Ok(BlockSyncResponse {
+            chain,
+            status,
+            message,
+        })
+    }
+}
+
+/// Contains the blocks and messages in a particular tipset
+#[derive(Clone, Debug, PartialEq)]
+pub struct TipSetBundle {
```

Comment:
Slightly confused with this type...could we not utilize the pre-existing `FullTipSet` which has a `Vec<Blocks>` that contains the bls and secp messages and would describe the block each message belongs to? 

Further, based on the spec and our existing implementation `blocks` should only contain a single header not a `Vec<BlockHeader>`?

Reply:
nah, this type defines specifically what is sent via the blocksync protocol. They use the tipsetbundle to remove redundant data passed over the wire. 

Also what type are you saying to only include one header? for even a single tipset there should always be a vector of headers (or blocks)

Reply:
I see that makes sense, although not completely clear how its a compressed Full Tipset as it contains the same data fields (e.g. blockheader fields and messages) and even adds an additional field with `msg_includes`? Could you not still use message vectors from a FullTipset?

Reply:
It's "compressed" because a tipset could have duplicate messages in the tipset in different blocks, the FullTipset is a vector of full blocks with messages, and the TipSetBundle has a dump of all headers, unique messages, and indexes for headers to messages